File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        coins = [0] + coins
        n = len(coins)
        dp = [[1e9]*(amount+1) for _ in range(n)]

        for i in range(1, n):
            dp[i][0] = 0
        
        for i in range(1, n):
            coin = coins[i]
            for j in range(1, amount+1):
                if j < coin:
                    dp[i][j] = dp[i-1][j]
                    continue

                dp[i][j] = min(dp[i-1][j-coin] + 1, dp[i][j-coin] + 1, dp[i-1][j])
        
        return dp[-1][-1] if dp[-1][-1] < 1e9 else -1

File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int n;
    int INF = 1e9;
    int answer;

    int dfs(vector<int>& coins, int amount, vector<int>& mem) {
        int min_cost = INF;

        if (amount == 0) {
            return 0;
        }

        if (mem[amount]) {
            return mem[amount];
        }

        for (int coin: coins) {
            if (amount-coin < 0) {
                continue;
            }
            
            min_cost = min(dfs(coins, amount-coin, mem)+1, min_cost);
        }

        mem[amount] = min_cost;
        return mem[amount];
    }


    int coinChange(vector<int>& coins, int amount) {
        n = coins.size();
        vector<int> mem(amount+1, 0);
        answer = dfs(coins, amount, mem);

        return answer == INF ? -1 : answer;
    }
};